### ISSUE TITLE
FileTraverser has a poor API for use by Kotlin

### ISSUE THREAD

**Issue by fzakaria** [google/guava#6137]
```
Overload resolution ambiguity. All these functions match.
public abstract fun breadthFirst(p0: Path!): (Mutable)Iterable<Path!>! defined in com.google.common.graph.Traverser
public abstract fun breadthFirst(p0: (Mutable)Iterable<Path!>!): (Mutable)Iterable<Path!>! defined in com.google.common.graph.Traverser
```

Kotlin cannot distinguish between Path and Iterable<Path> making use of this function impossible by Kotlin code.

**Comment by chaoren** [comment]
Are you talking about `MoreFiles.fileTraverser().breadthFirst()`? It works fine for me. Could you give some more details?

**Comment by fzakaria** [comment]
It doesn't show well from the snippet but what I pasted was the error the kotlin compiler was giving me.
It cannot distinguish between Iterable<Path> and Path since Path also implements the Iterable Interface.

Interestingly, I got around this by doing:
```
MoreFiles.fileTraverser().depthFirstPreOrder(listOf(root))
```
If I do the path without a collection `depthFirstPreOrder(root)` I get the failure as mentioned above.

**Comment by fzakaria** [comment]
This issue affects the depthFirstPreOrder and breadthFirst as they share the same disambiguity.

**Comment by chaoren** [comment]
What version of the Kotlin compiler are you using?

**Comment by chaoren** [comment]
Does the following example work for you?

https://pl.kotl.in/5FE0Jbj-Q

```
import java.nio.file.Path
import kotlin.io.path.Path

fun main() {
    foo(Path("."))
    foo(listOf(Path("/")))
}

fun foo(path: Path) {
    println(path)
}

fun foo(paths: Iterable<Path>) {
    println(paths)
}
```

**Comment by chaoren** [comment]
```
$ cat Test.kt
import com.google.common.io.MoreFiles
import java.nio.file.Path
import kotlin.io.path.Path

fun main() {
  for (path in MoreFiles.fileTraverser().breadthFirst(Path("."))) {
    println(path)
  }
}
$ wget https://repo1.maven.org/maven2/com/google/guava/guava/31.1-jre/guava-31.1-jre.jar
$ wget https://github.com/JetBrains/kotlin/releases/download/v1.7.10/kotlin-compiler-1.7.10.zip
$ unzip kotlin-compiler-1.7.10.zip
$ ./kotlinc/bin/kotlinc -cp guava-31.1-jre.jar:kotlinc/lib/kotlin-stdlib-jdk7.jar:. Test.kt
```

It compiles just fine.

```
$ ./kotlinc/bin/kotlin -cp guava-31.1-jre.jar:kotlinc/lib/kotlin-stdlib-jdk7.jar:. TestKt
```

It runs just fine as well.

**Comment by chaoren** [comment]
I'll go ahead and close this issue since it's not reproducible with the latest guava and kotlinc versions.

**Comment by michaelhixson** [comment]
Not sure what version OP is using, but I can reproduce the error locally in Kotlin 1.6.10 and 1.6.21, but not in 1.7.0 or latest, so it seems like they fixed it in the 1.7 release.

Using `breadthFirst(listOf(path))` seems like a reasonable workaround if updating Kotlin isn't viable yet.

**Comment by fzakaria** [comment]
Ah -- thanks for looking into it.
I will upgrade my Kotlin compiler.
